include/fmt/format.h

```diff
@@ -3162,48 +3309,66 @@ class dynamic_formatter {
     void on_hash() {}
   };
 
+  typedef internal::dynamic_format_specs<Char> dynamic_specs;
+
  public:
   template <typename ParseContext>
   auto parse(ParseContext &ctx) -> decltype(ctx.begin()) {
     auto it = internal::null_terminating_iterator<Char>(ctx);
+
+    typedef fmt::internal::dynamic_arg_ref_creator<ParseContext>
+        arg_ref_creator;
+    typedef fmt::internal::dynamic_specs_handler<dynamic_specs, ParseContext,
+                                                 arg_ref_creator>
+        handler_type;
+    arg_ref_creator creator(ctx);
+    handler_type handler(specs_, ctx, creator);
+
     // Checks are deferred to formatting time when the argument type is known.
-    internal::dynamic_specs_handler<ParseContext> handler(specs_, ctx);
     it = parse_format_specs(it, handler);
     return pointer_from(it);
   }
 
   template <typename T, typename FormatContext>
   auto format(const T &val, FormatContext &ctx) -> decltype(ctx.out()) {
-    handle_specs(ctx);
-    internal::specs_checker<null_handler>
-        checker(null_handler(), internal::get_type<FormatContext, T>::value);
-    checker.on_align(specs_.align());
-    if (specs_.flags == 0);  // Do nothing.
-    else if (specs_.has(SIGN_FLAG))
-      specs_.has(PLUS_FLAG) ? checker.on_plus() : checker.on_space();
-    else if (specs_.has(MINUS_FLAG))
+    basic_format_specs<Char> specs;
+    if (ctx.has_prepared_specs()) {
+      specs = ctx.get_prepared_specs();
```

Comment:
It seems strange to have two different specs here, `specs_` and the ones returned by `ctx.get_prepared_specs()`. Can we avoid that?

Ideally format_context shouldn't be aware of specs at all. Those should be managed by each individual formatter.

Reply:
> What do you think?

Wouldn't the second option suffer from the same problem?

In general I think the only way to communicate format specifiers should be a formatter object which keeps very clear separation of formatting state from the rest. At least that's the current design and I would very much try to keep it this way unless there are very strong reasons why it's impossible.

In this particular case, I think we should only use `dynamic_format_specs` for formatting and convert relocatable `prepared_format_specs` into them just before calling the `format` function.

Reply:
Probably it would but a new member would not be added to the `base_format_context`.

I see your point and I would not want to change the desing too.

Sure, specs can be converted before formatting but the problem remains. There is no way to pass prepared specs into a custom formatter. The formatter used to parse and format custom value is created deep in the `value::format_custom_arg()`. We'd need to create the formatter with/without prepared specs before and somehow pass it to the `value::format_custom_arg()`.
Another problem is that the type of the value is needed to determine the type of its formatter.

Reply:
Or maybe instead of passing whole formatter, we could pass only specs. That would solve the problem with figuring out the formatter type.

Reply:
Can we always use relocatable `prepared_format_specs` instead of `dynamic_format_specs` and avoid this problem in the first place? This would also simplify the code because we won't need two separate classes to represent specs for one narrow use case (variant-like types). Or am I still missing something?